### PR TITLE
3.x endpoints: Added payload to manifest order and added print endpoints

### PR DESCRIPTION
**Pull request by PatrickWaweru**
3.x endpoints: Added payload to manifest order and added print endpoints

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java [hunk 1]
```diff
@@ -187,6 +201,158 @@ public Object getValidOrdersForManifest(HttpServletRequest request, @RequestPara
...
+            System.out.println("Lab Manifest: The returned file was null");
+            return new ResponseEntity<String>("Lab Manifest: The returned file was null: ", HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        return(response);
+    }
+
+    /**
+     * Print the manifest given a lab manifest uuid
+     * To test use: curl -u admin:Admin123 -o specimen-label.pdf http://127.0.0.1:8080/openmrs/ws/rest/v1/kemrorder/printmanifest?manifestUuid=e50b861d-b7ce-49c1-91b7-4876de3abdf9
```
Comment (ojwanganto) [inline omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:253]: remove this comment

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java [hunk 2]
```diff
@@ -187,6 +201,158 @@ public Object getValidOrdersForManifest(HttpServletRequest request, @RequestPara
...
+            System.err.println("Lab Manifest: The returned file was null");
+            return new ResponseEntity<String>("Lab Manifest: The returned file was null", HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        return(response);
+    }
+
+    /**
+     * Print the manifest LOG given a lab manifest uuid
+     * To test use: curl -u admin:Admin123 -o specimen-label.pdf http://127.0.0.1:8080/openmrs/ws/rest/v1/kemrorder/printmanifestlog?manifestUuid=e50b861d-b7ce-49c1-91b7-4876de3abdf9
```
Comment (ojwanganto) [inline omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:311]: same to this.

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java [hunk 3]
```diff
@@ -187,6 +201,158 @@ public Object getValidOrdersForManifest(HttpServletRequest request, @RequestPara
...
+        } catch (Exception ex) {
+            System.err.println("Lab Manifest: Failed to generate manifest LOG printout" + ex.getMessage());
+            ex.printStackTrace();
+            return new ResponseEntity<String>("Lab Manifest: Failed to generate manifest LOG printout: " + ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        if (generatedLog != null) {
+            try {
+                InputStream is = new FileInputStream(generatedLog);
+                // response.setContentType(MediaType.APPLICATION_PDF);
```
Comment (ojwanganto) [inline omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:335]: let's remove this if it's not used

**Comment by ojwanganto** [comment]
Thanks, @PatrickWaweru. I have left some comments for you to address. 
Also, let's add a rule for a peer review before one can merge to master branch. It is very difficult to fix problems in an already merged code

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java [hunk 4]
```diff
@@ -165,7 +179,7 @@ public Object getValidOrdersForManifest(HttpServletRequest request, @RequestPara
             so.put("cccKdod", ccc);
 
             so.put("dateRequested", Utils.getSimpleDateFormat("dd-MM-yyyy").format(order.getDateCreated()));
-            so.put("payload", "");
+            so.put("payload", "test");
             so.put("hasProblem", load.get("hasProblem"));
```
[the remark this answers is not in the document]
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:183]: It is when an order has something missing which would make sending the manifest to the lab fail. e.g when patient has no CCC, regimen problems etc. It turns true alerting the user there is a problem before adding the order to the manifest. It doesnt prevent adding the order. It is to issue a warning.

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java (hunk 1 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:253]: Ok

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java (hunk 2 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:311]: Done

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java (hunk 3 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/controller/KenyaemrOrderRestController.java:335]: Done

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/resource/LabManifestOrderResource.java [hunk 5]
```diff
@@ -42,7 +56,37 @@ public LabManifestOrder newDelegate() {
     @Override
     public LabManifestOrder save(LabManifestOrder labManifestOrder) {
         System.out.println("Saving a new manifest order: " + labManifestOrder.toString());
```
[the remark this answers is not in the document]
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/resource/LabManifestOrderResource.java:58]: Ok

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/resource/LabManifestOrderResource.java [hunk 6]
```diff
@@ -42,7 +56,37 @@ public LabManifestOrder newDelegate() {
...
+        if (LabOrderDataExchange.getSystemType() == ModuleConstants.CHAI_SYSTEM) {
+            payloadGenerator = new ChaiSystemWebRequest();
+        } else if (LabOrderDataExchange.getSystemType() == ModuleConstants.LABWARE_SYSTEM) {
+            payloadGenerator = new LabwareSystemWebRequest();
+        } else if (LabOrderDataExchange.getSystemType() == ModuleConstants.EDARP_SYSTEM) {
+            payloadGenerator = new EdarpSystemWebRequest();
+        }
+
+        if (payloadGenerator == null) {
+            throw new ResourceDoesNotSupportOperationException("An error occured while adding sample to manifest: could not create payload generator: check system type");
```
[the remark this answers is not in the document]
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/resource/LabManifestOrderResource.java:76]: Let me check

[the remark this answers is not in the document]
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/web/resource/LabManifestOrderResource.java:76]: GenericRestException looks more appropriate

**Comment by PatrickWaweru** [comment]
> Thanks, @PatrickWaweru. I have left some comments for you to address. Also, let's add a rule for a peer review before one can merge to master branch. It is very difficult to fix problems in an already merged code

Ok. Let me fix the issues above and issue a new PR. I will also add the github rule.

**Pull request merged**